library/src/main/kotlin/com/freeletics/coredux/ReduxStore.kt

```diff
@@ -166,3 +166,32 @@ class ReducerException(
  * State update will always be received on [createStore] [CoroutineScope] thread.
  */
 typealias StateReceiver<S> = (S) -> Unit
+
+/**
+ * Filters [StateReceiver] by only passing states that are distinct from their predecessor state using [isStateDistinct]
+ * function.
+ *
+ * @param [isStateDistinct] receives `previousState` (can be null initially), `newState` and outputs `true` if new state
+ * should be passed to [StateReceiver] or `false` if not.
+ */
+fun <S : Any> StateReceiver<S>.distinctUntilChangedBy(
+    isStateDistinct: (previousState: S?, newState: S) -> Boolean
+): StateReceiver<S> {
+    var previousState: S? = null
+
+    return {
+        if (isStateDistinct(previousState, it)) {
+            previousState = it
+            this(it)
+        }
+    }
+}
+
+/**
+ * Filters [StateReceiver] by only passing states that are distinct from their predecessor state using states equality,
+ * only non-equal states will pass.
+ */
+fun <S : Any> StateReceiver<S>.distinctUntilChanged(): StateReceiver<S> =
```

Comment:
Shouldn't this be applied to a `StateReceiver` as soon as it subscribes?

Reply:
Isn't this always the case? I can't think about a specific case where you don't want to have distinct state changes.

Reply:
you may have special case of distinct that can be implemented using `distinctUntilChangedBy { .. }`

Point here that core should not force some approach and using distinct is optional. If it is optional, then either it is wrapper class around `StateReceiver<S>` or using this extension functions. Also for RxRedux we are using `.distinctUntilChanged()` chain operator as well.

Reply:
Isn't it possible to do a reference check on the state before and after calling the reducer? Then we know if the state has been changed.
My intention behind pulling `.distinctUntilChanged()` into the store is that it is possible to forget that call. So it is less error prone by pulling it into the store.